File: Clients/AndroidQ/QAnd/src/com/qframework/core/TextureFactory.java
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

package com.qframework.core;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Vector;

import javax.microedition.khronos.opengles.GL10;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;


import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.opengl.GLUtils;
import android.util.Log;

public class TextureFactory {

	public enum Type {
		DEFAULT,
		FONT,
	}

	public float mU1;
	public float mV1;
	public float mU2;
	public float mV2;
	public float mCp;

	public int mTextureDefault;
	public int mTextureFont;
	HashMap<String, Integer>	mTextures = new HashMap<String, Integer>();
	Vector<Integer>  mTextureIds = new Vector<Integer>();
	Vector < StringString > mInfos = new Vector <StringString>();
        Vector<String>	mToDelete = new Vector<String>();
	Context mContext;
	int mTextCounter = 1;
	boolean mUpdated = false;
	GameonApp mApp;
	public TextureFactory(GameonApp app)
	{
		mApp = app;
	}
	
	private int loadTextureFromFile(GL10 gl , String filename, Context context, boolean system)
	{
        int[] textures = new int[1];
        gl.glGenTextures(1, textures, 0);
        String text = " texture " + filename + " "  + mTextCounter + " " + textures[0];
        //Log.d("model", text);
        int textureid = mTextCounter ; //textures[0];
        mTextCounter++;
        //int textureid = textures[0];
        gl.glBindTexture(GL10.GL_TEXTURE_2D, textureid);
        gl.glTexParameterf(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR);
        gl.glTexParameterf(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR);
        gl.glTexParameterf(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_WRAP_S, GL10.GL_CLAMP_TO_EDGE);
        gl.glTexParameterf(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_WRAP_T, GL10.GL_CLAMP_TO_EDGE);
        mUpdated = true;
        gl.glTexEnvf(GL10.GL_TEXTURE_ENV, GL10.GL_TEXTURE_ENV_MODE,GL10.GL_MODULATE);
        AssetManager assets = context.getAssets();
	    
	    InputStream is = null;
		try {
			if (system)
			{
				is = assets.open( "qres/" + filename );				
			}else
			{
				is = assets.open( "res/" + filename );
			}
		} catch (IOException e1) {
			// TODO Auto-generated catch block
			e1.printStackTrace();
			return -1;
		}        
        Bitmap bitmap;
        try {
            bitmap = BitmapFactory.decodeStream(is);
        } finally {
            try {
                is.close();
            } catch(IOException e) {
                // Ignore.
            }
        }
        
        GLUtils.texImage2D(GL10.GL_TEXTURE_2D, 0, bitmap, 0);
        
        return textureid;

	}

	public void init(GL10 gl, Context context) {

		mContext = context;
		mTextCounter  = 1;
		mTextures.clear();
		mTextureDefault = loadTextureFromFile( gl, "whitesys.png" , mContext , true);
		if (mApp.cs().getCanvasW() < 500)
		{
			mTextureFont = loadTextureFromFile( gl, "fontsyss.png" , mContext , true);
		}
		else
		{
			mTextureFont = loadTextureFromFile( gl, "fontsys.png" , mContext, true);
		}
		for (int a=0; a< mInfos.size(); a++)
		{
			StringString e = mInfos.get(a);
			newTexture(gl,e.b, e.a , false);
		}

        //loadAssetsTextures(gl);
        
	}	
	public int get(Type type) {
		switch (type) {
		
			case FONT: return mTextureFont; 
			case DEFAULT: return mTextureDefault;
			//mTextureFont;
		}
		//Log.d("model", " get failed ");
		return mTextureDefault;
	}
	public int getTexture(String strData) {
		//Log.d("model" , " getTexture " + strData);
		
		if (mTextures.containsKey(strData))
		{
			//Log.d("model" , " returning " + mTextures.get(strData));
			return mTextures.get(strData);
		}
		Log.d("model", " get failed " + strData);
		return mTextureDefault;
	}
	
	public void deleteTexture(GL10 gl, String textname)
	{
		mToDelete.add(textname);
	}
	
	public void flushTextures(GL10 gl)
	{
		for (int a=0; a< mToDelete.size(); a++)
		{
			clearTexture(gl, mToDelete.get(a));
		}
		mToDelete.clear();
	}
	
	private void clearTexture(GL10 gl, String textname)
	{
		if (mTextures.containsKey(textname))
		{
			int id = mTextures.get(textname);
			gl.glDeleteTextures(1 , new int[]{id} ,0);
			mTextures.remove(textname);
			for (int a=0; a < mInfos.size(); a++)
			{
				StringString info = mInfos.get(a);
				if (info.b.equals(textname))
				{
					mInfos.remove(a);
					break;
				}
			}
		}
		
	}
	public void clear() {
		// TODO Auto-generated method stub
		mTextures.clear();
		mInfos.clear();
		mTextCounter = 1;
		
	}
	

	
	public void newTexture(GL10 gl,String textname , String textfile, boolean add)
	{
	
		int texture =  loadTextureFromFile( gl, textfile , mContext, false);

		if (texture > 0)
		{
			mTextures.put(textname, texture);
			if (add)
			{
				StringString info = new StringString();
				info.a = textfile;
				info.b = textname;
				mInfos.add( info );
			}
			if (textname.equals("font"))
			{
				mTextureFont = texture;
			}			
		}				
		
	}
	public boolean isUpdated()
	{
		return mUpdated;
	}
	
	public void resetUpdated()
	{
		mUpdated = false;
	}

	public  void setParam(float u1, float v1, float u2, float v2, float cp) {
		mU1 = u1;
		mV1 = v1;
		mU2 = u2;
		mV2 = v2;
		mCp = cp;
		
	}
	
    public void initTextures(GL10 gl, JSONObject response)
    {
        // init layout
		try {
	    	JSONArray areas;
			areas = response.getJSONArray("texture");
        
	        for (int a=0; a< areas.length(); a++)
	        {
	            JSONObject pCurr = areas.getJSONObject(a);
	            processTexture(gl, pCurr);
	        }
		} catch (JSONException e) {
			e.printStackTrace();
        }

	}

    private void processTexture(GL10 gl, JSONObject areaData) {
    	try {
			String name = areaData.getString("name");
			String file = areaData.getString("file");
			
			if (name != null && name.length() > 0 && file != null && file.length() > 0)
			{
				newTexture(gl ,name , file , true);
			}
    	}
    	catch (JSONException e) {
		
    		e.printStackTrace();
         }    	
    }
    
}

File: Clients/Html5Q/tutorials/qframework/TextureFactory.js
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/

function StringString()
{
	this.a = "";
	this.b = "";
}

	TextureFactory_Type =
	{
		DEFAULT:0,
		FONT:1
	}

function TextureFactory( app) 
{
    
	this.mTextureDefault = 1;
	this.mTextureFont = 1;
	this.mTextures = {};//new HashMap<String, Integer>();
	this.mTextureIds = []; //new Vector<Texture>();
	this.mInfos = []; //new Vector<StringString>();
	this.mToDelete = []; //new Vector<String>();
	this.mApp = app;
	this.mU1 = 0.0;
	this.mV1 = 0.0;
	this.mU2 = 0.0;
	this.mV2 = 0.0;
	this.mCp = 0.0;

}

TextureFactory.prototype.init = function(gl) {

	this.clear();
	this.mTextureDefault = this.loadTextureFromFile( gl, "whitesys.png" , true);
	this.mTextureFont = this.loadTextureFromFile( gl, "fontsys.png" , true);
	for (var a=0; a< this.mInfos.length; a++)
	{
		var e = this.mInfos.get(a);
		this.newTexture(gl,e.b, e.a , false);
	}
	
}	

TextureFactory.prototype.handleLoadedTexture = function(texture) {
	gl.bindTexture(gl.TEXTURE_2D, texture);
	//gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, true);
	gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, texture.image);
	gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
	gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
	gl.bindTexture(gl.TEXTURE_2D, null);
}
	
TextureFactory.prototype.loadTextureFromFile = function(gl , fname , system)
{
	//id = loadImageTexture(gl, "res/" + fname);
	var id = gl.createTexture();
	id.image = new Image();
	id.image.callback = this;
		id.image.onload = function () {
		this.callback.mApp.mDataChange = true;
		this.callback.handleLoadedTexture(id)
	}

	if (system == true)
	{
		id.image.src = "qres/" + fname;
	}
	else
	{
		id.image.src = "res/" + fname;
	}
	
	return id;
}

TextureFactory.prototype.get = function(type) {
	switch (type) {
	
		case TextureFactory_Type.FONT: return this.mTextureFont;
		case TextureFactory_Type.DEFAULT: return this.mTextureDefault;
	}
	return this.mTextureDefault;
}

TextureFactory.prototype.getTexture = function(strData) {

	if (this.mTextures[strData] != undefined)
	{
		//Log.d("model" , " returning " + mTextures.get(strData));
		return this.mTextures[strData];
	}
	return this.mTextureDefault;
}

TextureFactory.prototype.deleteTexture = function(gl ,textname)
{
	this.mToDelete.push(textname);
}

TextureFactory.prototype.flushTextures = function(gl)
{
	for (var a=0; a< this.mToDelete.length ; a++)
	{
		this.clearTexture(gl , this.mToDelete[a]);
	}
	
	this.mToDelete = [];
}

TextureFactory.prototype.clearTexture = function(gl, textname)
{
	if (this.mTextures[textname] != undefined)
	{
		var id = this.mTextures[textname];
		gl.deleteTexture(id);
		delete this.mTextures[textname];
		for (var a=0; a < this.mInfos.length; a++)
		{
			var info = this.mInfos[a];
			if (info.b == textname)
			{
				this.mInfos.splice(a,1);
				break;
			}
		}
	}
	
}

TextureFactory.prototype.clear = function() {
	// TODO Auto-generated method stub
	this.mTextures = {};
	this.mInfos = [];
			
}


TextureFactory.prototype.newTexture = function(gl,textname , textfile, add)
{
	
	
	var texture =  this.loadTextureFromFile( gl, textfile , false);
	if (texture != undefined)
	{
		this.mTextures[textname] = texture;
		if (add)
		{
			var info = new StringString();
			info.a = textfile;
			info.b = textname;
			this.mInfos.push( info );
		}
		if (textname== "font")
		{
			this.mTextureFont = texture;
		}		
	}else
	{
		//System.out.println("cant load " + textname);
	}
	
}

TextureFactory.prototype.setParam = function( u1,v1,u2,v2, cp)
{
	this.mU1 = u1;
	this.mU2 = u2;
	this.mV1 = v1;
	this.mV2 = v2;	
	this.mCp = cp;	
}


TextureFactory.prototype.initTextures = function(response)
{
	var areas;
	areas = response["texture"];
	for (var a=0; a< areas.length; a++)
	{
		var pCurr = areas[a];
		this.processTexture(pCurr);
	}	

}

TextureFactory.prototype.processTexture = function( objData) {
	var name = objData["name"];
	var file = objData["file"];
	this.newTexture(gl,name , file , true);
	

}

